llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,461 @@
+  auto CurrentPhysReg = VRM.getPhys(Def.getReg());
+  for (const auto &Reg : UndefPartReuse) {
+    if (Reg.first == ID)
+      continue;
+
+    if (VRM.getPhys(Reg.first) == CurrentPhysReg &&
+        reuseSubRegs(UndefPartReuse[ID], UndefPartReuse[Reg.first]))
+      return true;
+  }
+  return false;
+}
+
+static bool isNotLastDef(const MachineOperand &Def,
+                         const MachineInstr &DefInstr,
+                         const MachineBasicBlock &MBB) {
+  const Register Reg = Def.getReg();
+  bool FoundDef = false;
+  for (const MachineInstr &MI : MBB) {
+    if (!MI.definesRegister(Reg))
+      continue;
+    if (MI.isIdenticalTo(DefInstr)) {
+      FoundDef = true;
+      continue;
+    }
+    if (FoundDef)
+      return true;
+  }
+  return false;
+}
+
+static bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                         SmallVector<const MachineInstr *, 4> &OtherInstr) {
+  for (auto [InstrA, InstrB] : zip(OrigInstr, OtherInstr)) {
+    if (!InstrA->isCopyLike())
+      continue;
+    if (!InstrB->isCopyLike())
+      continue;
+
+    for (auto [OpA, OpB] : zip(InstrA->uses(), InstrB->uses())) {
+      if (!OpA.isReg() || !OpB.isReg())
+        continue;
+      if (OpA.getReg() != OpB.getReg())
+        continue;
+      if (OpA.getSubReg() != OpB.getSubReg())
+        continue;
+
+      if (containEqualSubRegDefs(InstrA, InstrB))
+        return true;
+    }
+  }
+  return false;
+}
+
+static bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                                   const MachineInstr *InstrB) {
+  for (auto [defA, defB] : zip(InstrA->defs(), InstrB->defs())) {
+    if (defA.getSubReg() == defB.getSubReg())
+      return true;
+  }
+  return false;
+}
+
+bool isIdentityCopy(const MachineInstr &MI, const VirtRegMap &VRM) {
+  if (!MI.isCopy())
+    return false;
+
+  Register DstPhyReg = MI.getOperand(0).getReg();
+  if (DstPhyReg.isVirtual())
+    DstPhyReg = VRM.getPhys(DstPhyReg);
+
+  Register SrcPhyReg = MI.getOperand(1).getReg();
+  if (SrcPhyReg.isVirtual())
+    SrcPhyReg = VRM.getPhys(SrcPhyReg);
+
+  return DstPhyReg == SrcPhyReg;
+}
+
+static void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                         const MCPhysReg &PhysReg,
+                         const MachineRegisterInfo &MRI,
+                         const AIEBaseRegisterInfo &TRI) {
+  LLVM_DEBUG(dbgs() << "Adding to Forbidden Regs ("
+                    << printReg(PhysReg, &TRI, 0, &MRI) << ") with alias:");
+  for (MCRegAliasIterator AI(MCRegister(PhysReg), &TRI, true); AI.isValid();
+       ++AI) {
+    ForbiddenPhysRegs.insert(*AI);
+    LLVM_DEBUG(dbgs() << printReg(*AI, &TRI, 0, &MRI) << " ");
+  }
+  LLVM_DEBUG(dbgs() << "\n");
+}
+
+} // end anonymous namespace
+
+char AIEWawRegRewriter::ID = 0;
+char &llvm::AIEWawRegRewriterID = AIEWawRegRewriter::ID;
+
+INITIALIZE_PASS(AIEWawRegRewriter, "aie-waw-reg-rewrite", "AIE waw-reg rewrite",
```

Comment:
nit: second parameter can be `DEBUG_TYPE`.

Reply:
fixed it